### ISSUE TITLE
cleanup pmmr_leaf snapshot files (from fast-sync)

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#1348]
Right now during fast-sync the node we ware sync'ing _from_ generates a "snapshot" of the pmmr_leaf file based on a rewound txhashset.
These get saved to the local filesystem and are never cleaned up.

Probably makes sense to clean these up, maybe as part of `check_compact()` that runs on a periodic basis.
We want to be careful that we don't `rm` anything _during_ a sync to avoid sending corrupted data across, but these never get used again once sync has completed. 

```
ls -Alt .grin/txhashset/output/
total 5932
-rw-r--r-- 1 root root 1989068 Aug 12 20:09 pmmr_data.bin
-rw-r--r-- 1 root root   16408 Aug 12 20:09 pmmr_leaf.bin
-rw-r--r-- 1 root root 3743904 Aug 12 20:09 pmmr_hash.bin
-rw-r--r-- 1 root root   16408 Aug 12 19:28 pmmr_leaf.bin.1084a7f2
-rw-r--r-- 1 root root   16408 Aug 11 12:25 pmmr_leaf.bin.0114b520
-rw-r--r-- 1 root root   16408 Aug  7 14:14 pmmr_leaf.bin.02cdd912
-rw-r--r-- 1 root root   16408 Aug  7 03:16 pmmr_leaf.bin.03437f9a
-rw-r--r-- 1 root root   16408 Aug  6 08:32 pmmr_leaf.bin.09a0b9f6
-rw-r--r-- 1 root root    8208 Jul 21 00:29 pmmr_leaf.bin.0a94f33f
-rw-r--r-- 1 root root    8208 Jul 20 02:20 pmmr_leaf.bin.1089dfac
-rw-r--r-- 1 root root    8208 Jul 19 20:45 pmmr_leaf.bin.089cb892
-rw-r--r-- 1 root root    8208 Jul 19 03:51 pmmr_leaf.bin.325aae69
-rw-r--r-- 1 root root    8208 Jul 18 16:42 pmmr_leaf.bin.149dc380
-rw-r--r-- 1 root root    8208 Jul 18 11:24 pmmr_leaf.bin.1d9ffe1e
-rw-r--r-- 1 root root    8208 Jul 18 09:04 pmmr_leaf.bin.242d170c
-rw-r--r-- 1 root root    8208 Jul 17 07:25 pmmr_leaf.bin.19125da8
-rw-r--r-- 1 root root    8208 Jul 16 23:08 pmmr_leaf.bin.09dd9ae5
-rw-r--r-- 1 root root    8208 Jul 16 21:42 pmmr_leaf.bin.02018c12
-rw-r--r-- 1 root root    8208 Jul 16 14:50 pmmr_leaf.bin.071de112
-rw-r--r-- 1 root root    8208 Jul 15 21:29 pmmr_leaf.bin.3ab0c1a3
-rw-r--r-- 1 root root    8208 Jul 15 21:20 pmmr_leaf.bin.435b67c9
-rw-r--r-- 1 root root    8120 Jul 13 07:14 pmmr_leaf.bin.233e096c
-rw-r--r-- 1 root root    8208 Jul 13 03:30 pmmr_leaf.bin.02425ec7
-rw-r--r-- 1 root root    4314 Jul 12 06:47 pmmr_leaf.bin.086e8860
-rw-r--r-- 1 root root    3080 Jul 11 20:29 pmmr_leaf.bin.01e35437
-rw-r--r-- 1 root root    3066 Jul 11 20:19 pmmr_leaf.bin.070b1cd3
-rw-r--r-- 1 root root    1744 Jul 11 07:57 pmmr_leaf.bin.07ff8ffc
-rw-r--r-- 1 root root    1246 Jul 11 05:03 pmmr_leaf.bin.065e7c1a
```

**Comment by haarts** [comment]
If someone what's to pick this up and doesn't know what `pmmr`s are read this:
- https://github.com/opentimestamps/opentimestamps-server/blob/master/doc/merkle-mountain-range.md
- https://github.com/mimblewimble/grin/blob/master/doc/merkle.md#proposed-merkle-structure

**Comment by eupn** [comment]
> We want to be careful that we don't rm anything during a sync to avoid sending corrupted data across, but these never get used again once sync has completed.

Some checks are already in-place, one is [here](https://github.com/mimblewimble/grin/blob/46051ee17458817d99e2190d20d75873cf9e225d/servers/src/common/adapters.rs#L519):
```rust
    // no compaction during sync or if we're in historical mode
    if self.archive_mode || self.sync_state.is_syncing() {
        return;
    }
```

And one more check should be added to `POST /v1/chain/compact` API [handler](https://github.com/mimblewimble/grin/blob/46051ee17458817d99e2190d20d75873cf9e225d/api/src/handlers.rs#L545) to make it safe to be called during sync.

Or we can make `txshashset` or `chain` to know that is in sync state and place a check there.

**Comment by rentenmark** [comment]
@antiochp I am looking at picking this up. I have the build working (MacOS) just trying to figure out where this aught to be. I found where it gets written [here](https://github.com/mimblewimble/grin/blob/master/store/src/pmmr.rs#L203) I wasn't sure if there was a general place where things get deleted or pruned but it looks like it could be a feature of the `PMMRBackend` as that is the writer? In `check_compact` we have a few data members on `NetToChainAdapter` but it doesn't have a `PMMRBackend`.

My proposed approach would be to make the `NetToChainAdapter` have a `PMMRBackend` or create a new one in `check_compact` and implement a new function `clean_synced` or similar on `PMMRBackned` that will clean up the rewind files. I also plan to write tests for all this.

I'm looking for any advice on direction, thank you.

**Comment by antiochp** [comment]
I think we can actually do something a lot simpler than what is being proposed here.

The backend writes to `pmmr_leaf.bin` during normal operation. And when a peer asks to sync from us we "rewind" and write a "snapshot" like `pmmr_leaf.bin.1084a7f2`. The details of why we need a snapshot are not that important here, but we basically want to send state across for a specific known block height.

The issue is we create one of these snapshot files _every_ time a peer asks to sync and the list of these files grows over time.

We just need something that will clean these up periodically.
We don't need to read the contents of these and we don't even really need to know what these files are for.
So I think we can safely avoid needing to do anything involving the `PMMRBackend`.

Basic steps involved - 
* List the contents of that dir, looking for `pmmr_leaf.bin.*` 
* Remove all snapshot files older than `x` mins? (We don't want to remove anything actually in use currently)

And I think we can just hook into `check_compact()` as this runs on a daily basis on all nodes already.

Does this make sense?

**Comment by ignopeverell** [comment]
@antiochp makes sense to me but can you remind me why we can't just send the latest and let the client do the rewind when it got that?

**Comment by antiochp** [comment]
@ignopeverell to rewind the output MMR you need to be able to "unspend" outputs and the current state of the MMR may no longer have those.

I could spend really old outputs years after they were created. They may then immediately get pruned/compacted (we prune based on the age of the outputs, not the age of the "spend").

**Comment by ignopeverell** [comment]
Ah right! I keep forgetting about this as well, thanks for the refresher!
